elastic/kibana
# [Time to Visualize] Remove Panels from URL

## Summary
Fixes https://github.com/elastic/kibana/issues/71499

This PR accomplishes a number of things:

~~**Note**: The following changes will only be visible If `dashboard.allowByValueEmbeddables` is set to true in `config/kibana.dev.yml`~~ **Edit:** This option is now set to true by default

1. **Shorter Urls**: Dashboard panel state storage in edit mode is no longer handled by the URL. In edit mode, the URL should be significantly shorter.
2. **Stickier State**: Since panel state is now stored in Session Storage, you should not lose unsaved state unless you close your tab. Navigating anywhere else in Kibana and back to your dashboard through any means  should properly restore your edits.
3. **Multiple Edits**: You can now have ongoing unsaved edits to multiple dashboards in the same session and flip between them. 
4. **Recently Accessed**: The recently accessed section now remembers if you were in view mode or edit mode on the dashboard, and if you were editing it, it restores the panels. New dashboards now also show up in the recently accessed section
5. **Edit Icon** The pencil icon in the dashboard listing page is now a shortcut that opens the selected dashboard in edit mode initially.
6. **Callout**: A small callout section will appear on the dashboard listing page. It lists dashboards with unsaved edits and provides options to discard the changes, or continue editing.
7. **Discard vs View / Edit:** Leaving edit mode no longer prompts the user to discard their changes. You can flip between edit and view mode freely, and there is now an explicit button for discarding your unsaved changes.

<img width="1008" alt="Screen Shot 2021-01-14 at 5 03 36 PM" src="https://user-images.githubusercontent.com/14276393/104654707-8b735880-568a-11eb-87af-774c696c9683.png">

![Edit -  view](https://user-images.githubusercontent.com/14276393/104966901-0badfc80-59b0-11eb-91de-6c5fcc7ff662.gif)

**To do**: Functional Tests are incoming. Required functional tests are linked in https://github.com/elastic/kibana/issues/80584

## Testing Ideas
This PR has a lot of UX, so it's important we test every single edge case we can think of. Some edge cases I can think of include:
- Start editing a dashboard and leave. _Any way_ you can get back to that dashboard should restore your unsaved edits, including: back button, Listing page link, Listing page unsaved callout link, recently accessed link. etc.
- Save an existing dashboard, make sure that the unsaved edits to that dashboard have been cleared.
- Save an existing dashboard as a new dashboard, and ensure the edits to the original dashboard have been cleared.
- Create a new dashboard, ensure its unsaved edits get saved. Save the new dashboard, ensure that the unsaved edits are cleared.
- Do a snapshot share, ensure that the state is properly put into the URL. Open this URL in a new tab, it should populate those unsaved changes.
- Do a snapshot share with a short URL. That should work the same way as the previous item.
- There are probably plenty of other edge cases


### Checklist

Delete any items that are not applicable to this PR.


### For maintainers
